src/server.c

```diff
@@ -5115,23 +5118,32 @@ void loadDataFromDisk(void) {
                 (float)(ustime()-start)/1000000);
 
             /* Restore the replication ID / offset from the RDB file. */
-            if ((server.masterhost ||
-                (server.cluster_enabled &&
-                nodeIsSlave(server.cluster->myself))) &&
-                rsi.repl_id_is_set &&
+            if (rsi.repl_id_is_set &&
                 rsi.repl_offset != -1 &&
                 /* Note that older implementations may save a repl_stream_db
                  * of -1 inside the RDB file in a wrong way, see more
                  * information in function rdbPopulateSaveInfo. */
                 rsi.repl_stream_db != -1)
             {
-                memcpy(server.replid,rsi.repl_id,sizeof(server.replid));
-                server.master_repl_offset = rsi.repl_offset;
-                /* If we are a slave, create a cached master from this
-                 * information, in order to allow partial resynchronizations
-                 * with masters. */
-                replicationCacheMasterUsingMyself();
-                selectDb(server.cached_master,rsi.repl_stream_db);
+                if (!iAmMaster()) {
+                    memcpy(server.replid,rsi.repl_id,sizeof(server.replid));
+                    server.master_repl_offset = rsi.repl_offset;
+                    /* If this is a replica, create a cached master from this
+                     * information, in order to allow partial resynchronizations
+                     * with masters. */
+                    replicationCacheMasterUsingMyself();
+                    selectDb(server.cached_master,rsi.repl_stream_db);
+                } else if (!server.rdb_expired_keys_last_load) {
+                    /* If this is a master, and no key expired when loading,
+                     * we can save the replication info as a secondary ID and offset,
+                     * in order to allow replicas to partial resynchronizations
+                     * with masters. */
+                    memcpy(server.replid2,rsi.repl_id,sizeof(server.replid));
+                    server.master_repl_offset = rsi.repl_offset;
+                    server.second_replid_offset = server.master_repl_offset+1;
+                    createReplicationBacklog();
+                    server.repl_no_slaves_since = time(NULL);
```

Comment:
maybe we can refactor this to be shorter and easier to read by utilizing `shiftReplicationId`?
i.e. we always write into the last replid, and then either replicationCacheMasterUsingMyself and selectDb, or shiftReplicationId and createReplicationBacklog etc.

Reply:
this is a very rare scenario, I prefer to take it special